### ISSUE TITLE
Schema snapshotting inconsistency 

### ISSUE THREAD

**Issue by terion-name** [directus/directus#11240]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

I've made some changes on server. Snapshotted schema, put it in my local env (that lacks some collections), tried to apply, that failed (as usual, but t is another story). Ok, I've dumped the db from server, imported it in local db. Run schema apply — got bunch of useless updates (this will be next ticket), but it applied. And then I've snapshotted schema again. And file differes.

It is full of such blocks:

<img width="418" alt="Снимок экрана 2022-01-24 в 19 50 30" src="https://user-images.githubusercontent.com/1060205/150836959-e6462cd2-c75f-4785-864e-81be0f725bfc.png">
 
Nothing changed here but order of fields.

Why is this a problem? Because schema file is supposed to be version controlled. And when one adds a collection for example, dumps the schema — and instead of having proper diff with change in collection one gets a huge mess on tens of. thousands changes when nothing can be understood or tracked. It is critical to force strict order in yaml render

### To Reproduce

Snapshot schema in different environments

I suppose that difference is OS-dependent, server runs in docker, locally — MacOS

### What version of Directus are you using?

9.4.3

### What version of Node.js are you using?

v16.13.1

### What database are you using?

mysql

### What browser are you using?

brave

### What operating system are you using?

node base docker image, macos

### How are you deploying Directus?

Docker, kubernetes, locally

**Comment by rijkvanzanten** [comment]
What were the differences in how you created the schema?

**Comment by terion-name** [comment]
> What were the differences in how you created the schema?

em... what can be? `npx directus schema snapshot schema.yml` both times, I don't know any other way

**Comment by rijkvanzanten** [comment]
You mentioned there were environmental changes between the two operations you ran right? Could you elaborate on those differences?

**Comment by terion-name** [comment]
@rijkvanzanten but it s written in issue:

> I suppose that difference is OS-dependent, server runs in docker, locally — MacOS

namely official node 16.13.1 docker image on server and mac os with 16.12.0 locally

**Comment by rijkvanzanten** [comment]
@terion-name Any other differences that would be important to be aware of? Did you use the same database for both? Same MySQL version in case of different databases?

Long story short, in order to be able to look into this, I need to be able to recreate the exact setup you're using so I can reproduce and resolve the issue. In order to most efficiently do that, I need as much information as possible on how to get to the point you are 👍🏻

**Comment by terion-name** [comment]
@rijkvanzanten I don't think so. Mysql 8 on both (different dbs of course, local and remote, but same version), node versions and os/containers described... Container is running in kubernetes, but this should not matter.

db should be irrelevant I think. It should be with os or node, cause key order in objects is never guaranteed in any language except php, so yaml renderer outputs keys in order the system gave them in `Object.keys()` call or how is it internally implemented. 

In short, [js-yaml](https://www.npmjs.com/package/js-yaml) has `sortKeys` option that should be used

**Comment by rijkvanzanten** [comment]
> db should be irrelevant I think

Very relevant. If you for example used sqlite on one and mysql on the other, we're working with two different database abstractions who might return results in different formats and orders and whatever. 

As for guaranteed order, sure, JavaScript technically doesn't have a guaranteed object order, but running the same bit of logic on the same version of node, using the same version of v8 will result in the same output, so differences there are also good to be aware of. 

Sure, adding in sorting on the js-yaml layer will be a fix for this, as it hardcodes the order in a known format. That being said, it won't explain and resolve why this is happening in the first place, and won't work for non-yaml file formats either, causing this issue to show up again when somebody uses something else

**Comment by rijkvanzanten** [comment]
> but we both understand that this should not be env-dependent =) and here are two different node versions, so maybe that's the case

Totally! Just curious what's causing it in the first place, as we might be able to come up with a proper solution further up the tree, rather than adding a fix all the way at the output level (where it writes to a file) 🙂

**Comment by terion-name** [comment]
> rather than adding a fix all the way at the output level (where it writes to a file) 🙂

but you literally have to. js can't guarantee stable key order. not in any case. it can change due to any reason, being environment, object creation flow, key assigning order etc (I will not be surprised if even cpu family can affect this due to internal memory addressing stuff). forcing order on output level is the proper solution in this specific case — because it is about output) think about this as a render function

**Comment by rijkvanzanten** [comment]
>  js can't guarantee stable key order. not in any case.

Do you have a source for that? 

Last I heard, [ES2015 updated the spec](https://www.stefanjudis.com/today-i-learned/property-order-is-predictable-in-javascript-objects-since-es2015/) around what types of objects and methods are predictable. [The 2022 language specification](https://tc39.es/ecma262/#sec-ordinaryownpropertykeys) also seems to indicate there's at least some predictability to be found when it comes to object property order.

Please be careful making claims like _you literally have to_ and _not in any case_. The only thing we know for sure is that you can't know things for sure.

**Comment by terion-name** [comment]
@rijkvanzanten The predictability you mention has nothing to do with stable key ordering required in this case.

<img width="245" alt="Снимок экрана 2022-01-25 в 11 54 05" src="https://user-images.githubusercontent.com/1060205/150954033-28533e1c-bcd9-4534-90d5-24aa3ca692f3.png">
↑ this is predictable but not stable and pointless for stable formatted output


Your schema output is in fact a render of data, that must be row-to-row the same with the same input data regardless of how this data was constructed internally. In other words render output of `{a: 1, b: 1}` and `{b: 1, a: 1}` should be the same.  Relying on internal order is not stable: you slightly change implementation in any place in future version, or some plugins kick in, or race condition appears, or db versions return output in different order, or any other unpredictable scenario — and your output format is ruined. So yes, you literally have to enforce specific key order in render stage to keep schema format stable and suitable for version control. This is not "bad design" if you are concerned about that, it is normal solution for this specific case: you strictly form the output in a function as the output format is the responsibility of this function

**Comment by cah4a** [comment]
Agree with @terion-name.
There are zillions of causes, and nobody can predict what changes could affect key ordering. 

Moreover, taking care of key ordering is surely not the responsibility of every chunk of code that is used for schema extraction. It breaks the robustness principle: Be conservative in what you do, be liberal in what you accept from others.

If env or database or whatever returns different key order it shouldn't affect conservative output, yeah?

**Comment by rijkvanzanten** [comment]
yep, i'm not arguing against using a sorted render at the output step, i just would like to  confirm this is caused by what the hypothesis is above, before we end up patching against symptoms instead of solving the root problems. Hence the wish for more information to be able to accurately and empirically reproduce this locally.

**Comment by rijkvanzanten** [comment]
@abdonrd what are we looking at in your screenshot? The schema snapshot logic in Directus doesn't rely on SQL exports 🤔

**Comment by abdonrd** [comment]
> @abdonrd what are we looking at in your screenshot? The schema snapshot logic in Directus doesn't rely on SQL exports 🤔

When I run the:

`npx directus schema snapshot ./schema.yaml`

I get this in the `collections.collection.schema.sql`:

<img width="615" alt="Screenshot 2022-02-02 at 22 29 01" src="https://user-images.githubusercontent.com/1007051/152240461-4a84541d-d065-4e4b-91c6-017965c9d4bf.png">

**Comment by tedsteen** [comment]
Does this mean that the order of my fields will be forced to the sort order of the schema export after an export/apply ?

**Comment by rijkvanzanten** [comment]
> Does this mean that the order of my fields will be forced to the sort order of the schema export after an export/apply ?

No. The fields in your app will still be sorted based on the `sort` value in the actual schema snapshot. 

This issue discusses a bug in which the structure of the yaml file itself would change, which in turn caused Directus to notice an update, even though the values didn't change. That was since resolved 👍🏻
